### ISSUE TITLE
API versioning OData doesn't works with custom ODataQueryOptions

### ISSUE THREAD

**Issue by vvaskiv-lohika-tix** [dotnet/aspnet-api-versioning#780]
Framework: netcoreapp3.1

I'm having an issue with OData API versioning. When I don't use ODataQueryOptions in the action method, I received that the API version is not supported.

Here example of OData controller configuration. In this case implemented custom ODataQueryOptionsSlim, it's lightweight replacement of ODataQueryOptions. 

```
    [ApiVersion("1.0")]
    [ApiVersion("2.0")]
    [Route("")]
    [ControllerName("orders")]
    public class OrdersController : ODataController
    {
        [HttpGet]
        public IActionResult Get(ODataQueryOptionsSlim<OrderTemp> options) =>
            Ok(new[] { new OrderTemp() { Id = 1, Customer = "Bill Mei" } });

        [HttpGet("{key:int}")]
        public IActionResult Get(int key, ODataQueryOptionsSlim<OrderTemp> options) =>
            Ok(new OrderTemp() { Id = key, Customer = "Bill Mei" });
    }
```

When I used the controller with ODataQueryOptionsSlim, I always received that API version not supported. **See screenshot.**

![image](https://user-images.githubusercontent.com/90458768/135320063-2fd92aff-1219-4358-8fd1-2c50f3ad0773.png)

When the change from ODataQueryOptionsSlim to ODataQueryOptions, everything works.

```
    [ApiVersion("1.0")]
    [ApiVersion("2.0")]
    [Route("")]
    [ControllerName("orders")]
    public class OrdersController : ODataController
    {
        [HttpGet]
        public IActionResult Get(ODataQueryOptions<OrderTemp> options) =>
            Ok(new[] { new OrderTemp() { Id = 1, Customer = "Bill Mei" } });

        [HttpGet("{key:int}")]
        public IActionResult Get(int key, ODataQueryOptions<OrderTemp> options) =>
            Ok(new OrderTemp() { Id = key, Customer = "Bill Mei" });
    }
```

![image](https://user-images.githubusercontent.com/90458768/135320655-e284889a-7413-4ee8-979a-90e99df4bc42.png)

How to resolve this issue, in order to ODataQueryOptionsSlim also works with API versioning?
Does API versioning depend on ODataQueryOptions?
Maybe exist any attributes for mark action to supporting API versioning without ODataQueryOptions?

**Comment by vvaskiv-lohika-tix** [comment]
**ODataQueryOptionsSlim** implements interface of **IODataQueryOptions**. 
Yes, it works without API Versioning. See screenshots.

![image](https://user-images.githubusercontent.com/90458768/135331288-4ec091db-831e-4c23-9676-a67d865dd712.png)

![image](https://user-images.githubusercontent.com/90458768/135331430-81f22c4b-796a-409c-8c6b-b0fcc579fe2b.png)

![image](https://user-images.githubusercontent.com/90458768/135331535-72b175cc-03d8-490b-ac1a-24dc639536b8.png)

Regarding model binder, we created it. And one thing which I noticed it's that model binder doesn't execute when route with API versioning, but when without API versioning it's executing. I'm a little bit confusing, why model binder doesn't invoke on route with the API version.

![image](https://user-images.githubusercontent.com/90458768/135331946-9d770639-9173-47a0-92c6-9cc4ef6509ba.png)

**Comment by commonsensesoftware** [comment]
I believe that you have something working. However, there is no such [official] interface as **IODataQueryOptions** as can be seen [here](https://github.com/OData/WebApi/blob/master/src/Microsoft.AspNet.OData.Shared/Query/ODataQueryOptions.cs) and [here](https://github.com/OData/WebApi/blob/master/src/Microsoft.AspNetCore.OData/Query/ODataQueryOptions.cs). You also shouldn't need `[FromQuery]` because your own model binder would handle that.

Do you have the world's simplest repro with the necessary bits? It would make an investigation much faster.

**Comment by vvaskiv-lohika-tix** [comment]
Ohhh. Regarding IODataQueryOptions, it's my mistake. Sorry. What do you mean by 'simplest repro with the necessary bits'? Are you about the sample repository?

**Comment by commonsensesoftware** [comment]
Do you have, or can you provide, a version of your code that demonstrates the problem and can be uploaded to this issue for further investigation?

**Comment by vvaskiv-lohika-tix** [comment]
I created a simple example, to view the issue which I mentioned. 

https://github.com/vaskiv99/odata-api-versioning

**Comment by commonsensesoftware** [comment]
I realize it's been quite some time. Thanks for the repo. I haven't run it - yet. API Versioning doesn't really care about this, but I seem to recall that the older OData routing logic may have been affected by it. That logic wasn't extensible so some of it had to be forked by API Versioning. It's entirely possible that something was missed or not updated.

You might want to bounce this against the forthcoming `6.0` release [here](../tree/main/examples/AspNetCore/OData). The whole implementation changes for OData 8.0 and a bunch of things API Versioning used to have to do in order to make it work has been removed. If that work for you, then this issue may not be worth pursuing further.

**Comment by commonsensesoftware** [comment]
`6.0` _Preview 3_ is available with support for OData 8.0. I have a feeling this might work now.

**Comment by commonsensesoftware** [comment]
`6.0` has officially been released. In terms of routing, if it works with vanilla OData 8.x+, then it should _just work_ with API Versioning. `6.0` will still support .NET Core 3.1 as well as .NET 6.0. Thank you for the repro, it was useful. If this has somehow not been solved, I'm happy reopen the issue and investigate further. Thanks.
